core/src/main/scala/org/apache/spark/scheduler/Pool.scala

```diff
@@ -119,4 +119,66 @@ private[spark] class Pool(
       parent.decreaseRunningTasks(taskNum)
     }
   }
+
+  // Update the number of slots considered available for each TaskSetManager whose ancestor
+  // in the tree is this pool
+  // For FAIR scheduling, slots are distributed among pools based on weights and minshare.
+  //   If a pool requires fewer slots than are available to it, the leftover slots are redistributed
+  //   to the remaining pools using the remaining pools' weights.
+  // For FIFO scheduling, the schedulable queue is iterated over in FIFO order,
+  //   giving each schedulable the remaining slots,
+  //   up to the number of remaining tasks for that schedulable.
+  override def updateAvailableSlots(numSlots: Float): Unit = {
+    schedulingMode match {
+      case SchedulingMode.FAIR =>
+        val queueCopy = new ConcurrentLinkedQueue[Schedulable](schedulableQueue)
```

Comment:
do we need a concurrent queue here? We just need a copy so any queue can work, right?

Reply:
no we don't. Forgot to change it to regular non thread safe structure. Done now.